tests/sample/atomic_instruction.c

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+// Whenever this sample program changes, bpf2c_tests will fail unless the
+// expected files in tests\bpf2c_tests\expected are updated. The following
+// script can be used to regenerate the expected files:
+//     generate_expected_bpf2c_output.ps1
+//
+// Usage:
+// .\scripts\generate_expected_bpf2c_output.ps1 <build_output_path>
+// Example:
+// .\scripts\generate_expected_bpf2c_output.ps1 .\x64\Debug\
+
+#include "bpf_helpers.h"
+
+SEC("maps")
+struct bpf_map_def map = {
+    .type = BPF_MAP_TYPE_ARRAY, .key_size = sizeof(uint32_t), .value_size = sizeof(uint64_t), .max_entries = 1};
+
+SEC("xdp_prog") int func(struct xdp_md* ctx)
+{
+    long key = 0;
+    long* count = bpf_map_lookup_elem(&map, &key);
+    if (count) {
+        __sync_fetch_and_add(count, 1);
```

Comment:
Is it possible to have this generate both the fetch and non-fetch version of the instructions? I want to make sure what we generate is correct for both cases.

Reply:
See:
https://github.com/dthaler/ebpf-docs/blob/update/isa/kernel.org/instruction-set.rst#152atomic-operations

![image](https://user-images.githubusercontent.com/20480683/214896886-c364636f-0ab7-49b5-9a72-f2fdca3b78d3.png)

Reply:
Thanks to @dylandreimerink for providing the answer:
>Dylan Reimerink
  [1 hour ago](https://cilium.slack.com/archives/C4XCTGYEM/p1674753252369159?thread_ts=1674751865.973939&cid=C4XCTGYEM)
By using the return value or not.
The following example emits with a FETCH:
> #include <linux/types.h>

>
> #include "bpf_helpers.h"

>
>__u64 add64_value = 1;

>
>SEC("raw_tp/sys_enter")
>int add(const void *ctx)
>{
>    __u64 b = __sync_fetch_and_add(&add64_value, 2);

>
>    return b;
>}

>
>atomic: file format elf64-bpf

>
>Disassembly of section raw_tp/sys_enter:

>
>0000000000000000 <add>:
>       0:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>       2:       61 16 00 00 00 00 00 00 r6 = *(u32 *)(r1 + 0)
>       3:       85 00 00 00 0e 00 00 00 call 14
>       4:       bf 01 00 00 00 00 00 00 r1 = r0
>       5:       b7 00 00 00 00 00 00 00 r0 = 0
>       6:       77 01 00 00 20 00 00 00 r1 >>= 32
>       7:       5d 61 04 00 00 00 00 00 if r1 != r6 goto +4 <LBB0_2>
>       8:       b7 00 00 00 02 00 00 00 r0 = 2
>       9:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>      11:       db 01 00 00 01 00 00 00 r0 = atomic_fetch_add((u64 *)(r1 + 0), r0)

>
>0000000000000060 <LBB0_2>:
>      12:       95 00 00 00 00 00 00 00 exit
>This one emits without a FETCH:
>#include <linux/types.h>

>
>#include "bpf_helpers.h"

>
>__u64 add64_value = 1;

>
>SEC("raw_tp/sys_enter")
>int add(const void *ctx)
>{
>    __sync_fetch_and_add(&add64_value, 2);

>
>    return 2;
>}

>
>atomic: file format elf64-bpf

>
>Disassembly of section raw_tp/sys_enter:

>
>0000000000000000 <add>:
>       0:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>       2:       b7 02 00 00 02 00 00 00 r2 = 2
>       3:       db 21 00 00 00 00 00 00 lock *(u64 *)(r1 + 0) += r2
>       4:       b7 00 00 00 02 00 00 00 r0 = 2
>       5:       95 00 00 00 00 00 00 00 exit

Reply:
1. updated at https://github.com/microsoft/ebpf-for-windows/pull/1936/commits/731be45e5d2abaa8888c56f03b00a09e26e9790b
2. clang-11 doesn't recognize atomic instructions other than xadd, i use clang-15.0.6 to build atomic_instruction.c, and i notice we use clang-11.0.1 across this project, how do we handle this issue.
3. external\ebpf-verifier doesn't support atomic instruction yet, i temporarily add --no-verify in 
   user_mode_bpf2c.vcxproj and kernel_mode_bpf2c.vcxproj to make the build not fail.
4. seems even with clang15, when generating 32bit atomic instruction, only `__sync_fetch_and_add` works, other atomic functions like `__sync_fetch_and_sub` fail.

```
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u32 aa = 7;
    __sync_fetch_and_sub(&aa, 2);
    return 0;
}
PS > clang -target bpf -c .\atomic.c
fatal error: error in backend: Cannot select: 0x1e805cfd678: i64,ch = AtomicLoadSub<(load store seq_cst (s32) on
      %ir.3)> 0x1e805cfd610, FrameIndex:i64<1>, Constant:i64<2>
  0x1e805cfd268: i64 = FrameIndex<1>
  0x1e805cfd408: i64 = Constant<2>
In function: test
[...]
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u32 aa = 7;
    __sync_fetch_and_or(&aa, 2);
    return 0;
}
PS > clang -target bpf -c .\atomic.c
.\atomic.c:4:5: error: Unsupported atomic operations, please use 64 bit version
int test( void *ctx)
    ^
fatal error: error in backend: Cannot select: 0x1c474dc5f28: i64,ch = AtomicLoadOr<(load store seq_cst (s32) on       
      %ir.3)> 0x1c474dc5ec0, FrameIndex:i64<1>, Constant:i64<2>
  0x1c474dc5b18: i64 = FrameIndex<1>
  0x1c474dc5cb8: i64 = Constant<2>
In function: test

```

Reply:
Thanks for working on this and the great info!

> 2. clang-11 doesn't recognize atomic instructions other than xadd, i use clang-15.0.6 to build atomic_instruction.c, and i notice we use clang-11.0.1 across this project, how do we handle this issue.

For now I would just have tests that just use the byte code directly rather than compiling from C.  If compiling from C is absolutely needed for some reason, then maybe using asm() in the C to embed the unsupported ebpf instruction.  Using a later version of clang is tracked by issue #1111.

> 3. external\ebpf-verifier doesn't support atomic instruction yet, i temporarily add --no-verify in
>    user_mode_bpf2c.vcxproj and kernel_mode_bpf2c.vcxproj to make the build not fail.

Correct the verifier doesn't have support yet.  I would suggest holding off on bpf2c testing of atomic instructions until the verifier does support it.  Fine to manually test like you did but have atomics be skipped in cicd (I don't think we should let bpf2c support no-verify).

Reply:
Inline assembly is not recognized by clang11 either, shoud we leave only `__sync_fetch_and_add` in atomic_instruction.c
or just don't include test sample in this pr.

```
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u64 a = 2, b = 3, *c = &a;
    asm volatile("lock *(u64 *)(%0+0) += %1" : "=r"(c) : "r"(b), "0"(c));
    asm volatile("lock *(u64 *)(%0+0) &= %1" : "=r"(c) : "r"(b), "0"(c));
    return a;
}
PS > clang -target bpf -c .\atomic.c # clang15
PS > & 'c:\Program Files\LLVM-11\bin\clang.exe'  -target bpf -c .\atomic.c #clang11
.\atomic.c:8:18: error: invalid operand for instruction
    asm volatile("lock *(u64 *)(%0+0) &= %1" : "=r"(c) : "r"(b), "0"(c));
                 ^
<inline asm>:1:22: note: instantiated into assembly here
        lock *(u64 *)(r2+0) &= r1
                            ^
1 error generated.
```

Reply:
> shoud we leave only __sync_fetch_and_add in atomic_instruction.c or just don't include test sample in this pr?

Leave only the atomic add for things compiled from C.
Other atomic tests can act on bytecode generated directly from tests rather than by clang.

Reply:
8693960ac5ff: Split the atomic test into two test cases

Current ci was failed because i changed count increment in `tests/sample/droppacket.c` to atomic add and unit_test failed because [ubpf not support atomic instructions](https://github.com/iovisor/ubpf/issues/64) yet, should i remove this change or find a workaround?
https://github.com/microsoft/ebpf-for-windows/blob/8693960ac5ff0ef574959d5efd8bd05fa7fc3044/tests/sample/droppacket.c#L71-L72
Are there any other things I missed or did not do right?

Reply:
And for reference, here is the atomic_instruction_others.o disassembly output (llvm-objdump.exe LLVM version 15.0.6):
```
PS > llvm-objdump.exe -d ebpf-for-windows\tests\sample\atomic_instruction_others.o

ebpf-for-windows\tests\sample\atomic_instruction_others.o:     file format elf64-bpf

Disassembly of section xdp_prog:

0000000000000000 <test>:
       0:       7b 1a f8 ff 00 00 00 00 *(u64 *)(r10 - 8) = r1
       1:       b7 02 00 00 7b 00 00 00 r2 = 123
       2:       7b 2a f0 ff 00 00 00 00 *(u64 *)(r10 - 16) = r2
       3:       b7 00 00 00 03 00 00 00 r0 = 3
       4:       7b 0a e8 ff 00 00 00 00 *(u64 *)(r10 - 24) = r0
       5:       b7 01 00 00 00 00 00 00 r1 = 0
       6:       7b 1a c8 ff 00 00 00 00 *(u64 *)(r10 - 56) = r1
       7:       7b 1a e0 ff 00 00 00 00 *(u64 *)(r10 - 32) = r1
       8:       63 2a dc ff 00 00 00 00 *(u32 *)(r10 - 36) = r2
       9:       63 0a d8 ff 00 00 00 00 *(u32 *)(r10 - 40) = r0
      10:       63 1a d4 ff 00 00 00 00 *(u32 *)(r10 - 44) = r1
      11:       b7 01 00 00 01 00 00 00 r1 = 1
      12:       bf 12 00 00 00 00 00 00 r2 = r1
      13:       db 2a f0 ff 01 00 00 00 r2 = atomic_fetch_add((u64 *)(r10 - 16), r2)
      14:       b7 02 00 00 02 00 00 00 r2 = 2
      15:       db 2a f0 ff 41 00 00 00 r2 = atomic_fetch_or((u64 *)(r10 - 16), r2)
      16:       bf 02 00 00 00 00 00 00 r2 = r0
      17:       db 2a f0 ff 51 00 00 00 r2 = atomic_fetch_and((u64 *)(r10 - 16), r2)
      18:       b7 02 00 00 04 00 00 00 r2 = 4
      19:       db 2a f0 ff a0 00 00 00 lock *(u64 *)(r10 - 16) ^= r2
      20:       c3 1a dc ff 40 00 00 00 <unknown>
      21:       79 a1 e8 ff 00 00 00 00 r1 = *(u64 *)(r10 - 24)
      22:       db 1a f0 ff e1 00 00 00 r1 = xchg_64(r10 - 16, r1)
      23:       7b 1a e0 ff 00 00 00 00 *(u64 *)(r10 - 32) = r1
      24:       b7 01 00 00 05 00 00 00 r1 = 5
      25:       db 1a f0 ff f1 00 00 00 r0 = cmpxchg_64(r10 - 16, r0, r1)
      26:       79 a0 c8 ff 00 00 00 00 r0 = *(u64 *)(r10 - 56)
      27:       95 00 00 00 00 00 00 00 exit
```